### ISSUE TITLE
Pressure sensitivity stopped worked with wizardpen driver

### ISSUE THREAD

**Issue by loentar** [mypaint/mypaint#154]
I have UGEE M708 tablet (Wizardpen driver, detected as UC-LOIC TABLET 1060) which works perfectly on MyPaint 1.1.0 installed from Ubuntu 14.04 repo.

But in current version of MyPaint (built from latest git sources) pressure sensitivity does not work and the tablet is detected as mouse.

Also it have one more strange bug: when I press any hardware button on the tablet, current version MyPaint show About screen. I don't have hotkey for that!
In case of MyPaint 1.1.0 hardware buttons works correctly: switches between brushes, etc.

I suspect the problem may be when MyPaint picks for wrong device of three, but I cannot set other devices as "ignored" because mypaint will change the all three "UC-LOIC TABLET 1060" to ignored if I change any.

I can provide any required information.

```
$ xinput list
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ Barcode Reader                          	id=10	[slave  pointer  (2)]
⎜   ↳ UC-LOIC TABLET 1060                     	id=12	[slave  pointer  (2)]
⎜   ↳ UC-LOIC TABLET 1060                     	id=13	[slave  pointer  (2)]
⎜   ↳ UC-LOIC TABLET 1060                     	id=14	[slave  pointer  (2)]
⎜   ↳ SynPS/2 Synaptics TouchPad              	id=16	[slave  pointer  (2)]
⎣ Virtual core keyboard                   	id=3	[master keyboard (2)]
    ↳ Virtual core XTEST keyboard             	id=5	[slave  keyboard (3)]
    ↳ Power Button                            	id=6	[slave  keyboard (3)]
    ↳ Video Bus                               	id=7	[slave  keyboard (3)]
    ↳ Sleep Button                            	id=8	[slave  keyboard (3)]
    ↳ Barcode Reader                          	id=9	[slave  keyboard (3)]
    ↳ USB 2.0 Camera                          	id=11	[slave  keyboard (3)]
    ↳ AT Translated Set 2 keyboard            	id=15	[slave  keyboard (3)]

# pressure test
$ xinput test 12
.....
motion a[0]=1634 a[1]=1150 a[2]=0 
motion a[0]=1634 a[1]=1149 a[2]=0 
motion a[0]=1634 a[1]=1148 a[2]=0 
motion a[0]=1634 a[1]=1147 a[2]=0 
motion a[0]=1634 a[1]=1146 a[2]=0 
button press   1 a[0]=1634 a[1]=1146 a[2]=0 
motion a[0]=1634 a[1]=1146 a[2]=107 
motion a[0]=1634 a[1]=1146 a[2]=138 
motion a[0]=1634 a[1]=1146 a[2]=174 
motion a[0]=1634 a[1]=1146 a[2]=213 
motion a[0]=1634 a[1]=1146 a[2]=264 
motion a[0]=1634 a[1]=1146 a[2]=320 
motion a[0]=1634 a[1]=1146 a[2]=380 
motion a[0]=1634 a[1]=1146 a[2]=433 
motion a[0]=1634 a[1]=1146 a[2]=482 
motion a[0]=1634 a[1]=1146 a[2]=531 
motion a[0]=1634 a[1]=1146 a[2]=586 
.....
```

**Comment by achadwick** [comment]
Thanks for the screencast. Can you tell me what `xinput text-xi2` says for the device which once reported pressure (just the blurb it prints before any events are received).

* xinput doesn't report a pressure axis for id=14 above, but it does in your "pressure test" in the original report, so I really don't know where to start.
  * Does this tablet have a mouse/puck/ that came with it? Was that it? That's a very mouse-like device.
* Gdk isn't receiving pressure for this tablet according to the traces produced by https://gist.github.com/achadwick/54e5b111e5ac908becf1 . I'd expect to see something like

  ![Wacom Intuos5 PTH-650 with axischart](https://cloud.githubusercontent.com/assets/61299/5607551/996484e2-9457-11e4-9880-a13d4ed0dfd7.png)

  * Basically we can't be expected to do anything if we don't receive pressure. It is not MyPaint's job to manufacture pressure from other axes.

**Comment by loentar** [comment]
The tablet have no mouse, It just have a pen (with 2 buttons) and 8 "express keys" on the tablet. Here is it: http://ugee.net/products_view.asp?articleid=28 .

Additional info from xinput. Please note, device ids changed after I reboot my laptop:

```

$ xinput list

...
⎜   ↳ UC-LOIC TABLET 1060                     	id=11	[slave  pointer  (2)] --- tablet device
⎜   ↳ UC-LOIC TABLET 1060                     	id=12	[slave  pointer  (2)] --- ???
⎜   ↳ UC-LOIC TABLET 1060                     	id=13	[slave  pointer  (2)] --- tablet's keyboard
...

$ xinput test-xi2 11
UC-LOIC TABLET 1060                     	id=11	[slave  pointer  (2)]
	Reporting 4 classes:
		Class originated from: 11. Type: XIButtonClass
		Buttons supported: 6
		Button labels: None None None None None None
		Button state:
		Class originated from: 11. Type: XIValuatorClass
		Detail for Valuator 0:
		  Label: None
		  Range: 0.000000 - 3180.000000
		  Resolution: 1000 units/m
		  Mode: absolute
		  Current value: 1676.000000
		Class originated from: 11. Type: XIValuatorClass
		Detail for Valuator 1:
		  Label: None
		  Range: 0.000000 - 2048.000000
		  Resolution: 1000 units/m
		  Mode: absolute
		  Current value: 865.000000
		Class originated from: 11. Type: XIValuatorClass
		Detail for Valuator 2:
		  Label: None
		  Range: 0.000000 - 1023.000000
		  Resolution: 1000 units/m
		  Mode: absolute
		  Current value: 0.000000


# ... dropped lots of motion events

#... here I pressed the tablet using pen:

EVENT type 6 (Motion)
    device: 11 (11)
    detail: 0
    flags: 
    root: 1401.61/471.77
    event: 121.61/163.77
    buttons: 1
    modifiers: locked 0x10 latched 0 base 0 effective: 0x10
    group: locked 0 latched 0 base 0 effective: 0
    valuators:
        0: 1685.00
        1: 944.00
        2: 724.00                <<<<< this value changes on pressure change
    windows: root 0x9e event 0x4600001 child 0x0
EVENT type 17 (RawMotion)
    device: 11 (11)
    detail: 0
    valuators:
    flags: 
          0: 1685.00 (1685.00)
          1: 944.00 (944.00)
          2: 530.00 (530.00)     <<<<< this value changes on pressure change


# ============================================================

```

I made some tests with xinput and seems to be the second device #12 (???) is not used. If I disable it with `xinput disable 12` the tablet is working as usual, the pen's buttons/pressure works as well (on old MyPaint).

The most important thing for me: the tablet is detected as a mouse and not as a tablet in MyPaint. But why?

I tried to disable all other devices, including mouse (which is recognized as "Barcode Reader") and laptop's touchpad but nothing changes.

**Comment by loentar** [comment]
I'm see how gtk events works with my tablet using https://gist.github.com/achadwick/54e5b111e5ac908becf1.

I set a breakpoint at `_event_cb` and see the events in debug mode.

I see that all the events from tablet, mouse and laptop's touchpad are coming from "Virtual core pointer" device: https://drive.google.com/file/d/0BwvdX1pBENi1RjA4TWVrWnVBU00/view

But by some reason `value = event.get_axis(use)` always returns `None` for  `Gdk.AxisUse.PRESSURE`.

```
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ UC-LOIC TABLET 1060                     	id=11	[slave  pointer  (2)]
⎜   ↳ SynPS/2 Synaptics TouchPad              	id=16	[slave  pointer  (2)]
⎜   ↳ UC-LOIC TABLET 1060                     	id=12	[slave  pointer  (2)]
⎜   ↳ UC-LOIC TABLET 1060                     	id=13	[slave  pointer  (2)]
⎜   ↳ Barcode Reader                          	id=10	[slave  pointer  (2)]
```

when starting `xinput test-xi2 2` produces events for all pointer devices including tablet. Pressure info is available when using pen:

```
Virtual core pointer                    	id=2	[master pointer  (3)]
	Reporting 8 classes:
		Class originated from: 10. Type: XIButtonClass
		Buttons supported: 13
		Button labels: "Button Left" "Button Middle" "Button Right" "Button Wheel Up" "Button Wheel Down" "Button Horiz Wheel Left" "Button Horiz Wheel Right" "Button Side" "Button Extra" "Button Unknown" "Button Unknown" "Button Unknown" "Button Unknown"
		Button state:
		Class originated from: 10. Type: XIValuatorClass
		Detail for Valuator 0:
		  Label: Rel X
		  Range: -1.000000 - -1.000000
		  Resolution: 1 units/m
		  Mode: relative
		Class originated from: 10. Type: XIValuatorClass
		Detail for Valuator 1:
		  Label: Rel Y
		  Range: -1.000000 - -1.000000
		  Resolution: 1 units/m
		  Mode: relative
		Class originated from: 10. Type: XIValuatorClass
		Detail for Valuator 2:
		  Label: Rel Horiz Wheel
		  Range: -1.000000 - -1.000000
		  Resolution: 1 units/m
		  Mode: relative
		Class originated from: 10. Type: XIValuatorClass
		Detail for Valuator 3:
		  Label: Rel Vert Wheel
		  Range: -1.000000 - -1.000000
		  Resolution: 1 units/m
		  Mode: relative
		Class originated from: 10. Type: XIScrollClass
		Scroll info for Valuator 2
		  type: 2 (horizontal)
		  increment: 1.000000
		  flags: 0x0
		Class originated from: 10. Type: XIScrollClass
		Scroll info for Valuator 3
		  type: 1 (vertical)
		  increment: -1.000000
		  flags: 0x2 ( preferred )
		Class originated from: 16. Type: XITouchClass
		Touch mode: dependent
		Max number of touches: 2


# produces events with pressure info:

EVENT type 6 (Motion)
    device: 2 (11)
    detail: 0
    flags: 
    root: 646.32/548.23
    event: 597.32/524.23
    buttons: 1
    modifiers: locked 0x10 latched 0 base 0x4 effective: 0x14
    group: locked 0 latched 0 base 0 effective: 0
    valuators:
        0: 777.00
        1: 1097.00
        2: 810.00                             <<< pressure
    windows: root 0x9e event 0x6800001 child 0x0
EVENT type 17 (RawMotion)
    device: 2 (11)
    detail: 0
    valuators:
    flags: 
          0: 777.00 (777.00)
          1: 1097.00 (1097.00)
          2: 806.00 (806.00)              <<< pressure

```

Any thoughts why event does not have pressure info and device recognized as a mouse?

**Comment by achadwick** [comment]
I'm not that familiar with the guts of GDK, to be honest, though it looks like the problem lies there. You may want to look at differences between

* https://git.gnome.org/browse/gtk+/tree/gdk/x11/gdkinput-x11.c?h=gtk-2-24#n95
* https://git.gnome.org/browse/gtk+/tree/gdk/x11/gdkdevicemanager-xi2.c?h=gtk-3-14#n303

which are where the type field and axes get created for a GDK device object for GTK2 and GTK3. In other words, it's the ultimate source of the flags received for events.

I can see two immediate problems for GTK3 with this device, as xi2 reports it:

1. The name doesn't have "pen" in it ☺ which means that GDK_SOURCE_PEN won't be assigned to it. Not necessarily an issue, because mice can have pressure axes too, but:
2. The pressure axis is unlabelled on the "slave" (hardware) device, and (thus?) mislabelled as "Rel Horiz Wheel" in xinput2 on the corresponding "master" (virtual aggregate) device. Looks like GTK3-era GDK uses another string match (in [translate_valuator_class()](https://git.gnome.org/browse/gtk+/tree/gdk/x11/gdkdevicemanager-xi2.c?h=gtk-3-14#n155) to assign xi valuators to GDK axes, but GTK2-era GDK did... something different based on order?

It looks very much as though there's been a regression with the handling of this particular device, and you couldn't do worse than to ask the GTK and GDK folks about it on IRC or their mailing lists. I say that because it seems to be due to the way xinput/xi2 is reporting those axis labels, and because something is wrong there. They may suggest a udev-level workaround if you're lucky, and assuming axes can be assigned labels there. That sort of low-level stuff is beyond me ☺.

Here's a quick overview of the lower levels of Linux input device hotplug stuff: https://joeshaw.org/2010/10/01/681/ . All the stuff MyPaint deals with is beyond the end of the list given there :/

Sorry about this. All I can really do is tell you that it's Not MyPaint's Fault here, and where to start looking next.

**Comment by loentar** [comment]
I understand, it's not MyPaint failure.
Thank you very much for your help and suggestions. I will try it.

**Comment by loentar** [comment]
This note is for those who faced with the same problem and want for workaround.

The trouble is: the event does not contain information about axis type, but contain pressure info.

Workaround is: set third axis use to pressure in case of unsuccessful getting the pressure info and device has more than 2 axes.

Patch [is here](https://gist.github.com/loentar/da5ffce997fb3613db5d)

**Comment by achadwick** [comment]
Ah, so newer GDK is actually seeing the axis's value! Just flagged as `GDK_AXIS_IGNORE`. That's interesting. Good lateral thinking and a cool workaround :+1:

I'm in two minds about what we should do about this. Since the tablet in question and/or its lower-level input handling is fundamentally broken here I'm not going to chase it further. However I might very well accept a patch which allows users to remap axes by number, if it was complete with good UI and fitted into our [`gui.device.Monitor` framework](https://github.com/mypaint/mypaint/blob/599577b43e20a8f1a1c07113a20207af841062a4/gui/device.py#L70).
